Add repos description fetcher to generate pages

Crola1702 opened the pull request:
# Description

This PR adds `fetch_repositories.rb` script to fill `/content/repos` folder with all infra managed repositories.

The script consumes the repositories listed in `repos.yml` and fetches the information from GitHub API.

Crola1702 committed:
Add repos readme fetch to generate pages

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Put repo fetcher login outside nanoc

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Remove commented code

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Move fetcher instructions to README

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Update repos

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 commented:
@nuclearsandwich this is ready for review

Crola1702 committed:
Remove kramdown filter for repos

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

nuclearsandwich commented on Rules:
```diff
@@ -28,6 +31,12 @@ compile '/team/*.md' do
+compile '/repos/*.md' do
+  layout '/repo.erb'
+  layout '/default.erb'
+  write item.identifier.without_ext + '/index.html'
```
```suggestion
  write nil
```

I don't actually think we want to generate full pages per repository. That would get hard to manage quickly. I think that we can instead generate the content but not write it out, then use the generated item / content for the index page.

nuclearsandwich commented on fetch_repositories.rb:
```diff
@@ -0,0 +1,37 @@
+
+def write_repos(repos)
+  repos.each do |repo|
+    repo['name'] = repo['full_name'].gsub('/', '-')
```
I don't think there are any existing cases but this could create a collision if, e.g. we had projects called `ros-infrastructure/repo` and `ros-infrastructure/repo`. There's only marginally less risk if we use a single underscore as a separator so let's try two dashes.

```suggestion
    repo['name'] = repo['full_name'].gsub('/', '--')
```

Crola1702 replied:
I applied the suggestion and also updated the existing repos content. Check 9d0eef7

nuclearsandwich commented on layouts/repo.erb:
```diff
@@ -0,0 +1,7 @@
+<div class="row">
+  <h1>Repo: <a href="<%= @item[:url] %>"><%= @item[:name] %></a></h1>
+  <h5>Owner: <%= @item[:owner] %></h5>
```
Since the repository name includes the owner, I don't think we need this line at all.
```suggestion
```

nuclearsandwich commented on layouts/repo.erb:
```diff
@@ -0,0 +1,7 @@
+<div class="row">
+  <h1>Repo: <a href="<%= @item[:url] %>"><%= @item[:name] %></a></h1>
```
Since each repo is a repo, the prefix on the title seems redundant. If we were ever to print repositories alongisde other items, like collections of repositories in subprojects, it could make sense in the future.

```suggestion
  <h1><a href="<%= @item[:url] %>"><%= @item[:name] %></a></h1>
```

nuclearsandwich commented on layouts/repo.erb:
```diff
@@ -0,0 +1,7 @@
+  <h5>Owner: <%= @item[:owner] %></h5>
+  <br>
+  <h5>Description:</h5>
+  <p><%= @item[:description] || 'No description provided' %></p>
```
That this is the repository description will be clear from context and by omitting the heading we can also avoid printing "no description" when one is missing. 

```suggestion
  <% if @item[:description:] -%>
  <p><%= @item[:description] %></p>
  <% end -%>
```

nuclearsandwich committed:
Embed each repo on the repositories index.

The repos stored in the index attributes aren't usable as-is. I ended up
having to pull them out of `@items.find_all` while doing some digging I
discovered that while find_all is a linear time search, it does also
hint the dependency tracking system of nanoc and so it's not that
harmful to use (except in our very large case with the answers archive)

In order to get them to render correctly I also removed the default
layout from the repo item rule.

nuclearsandwich reviewed (changes requested):
I ended up pushing a change directly which is companion to some of these suggestions.

Crola1702 committed:
Add website content (#5)

Crola1702 committed:
Apply code suggestions

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Find all repositories once

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Remove repos preprocess

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Change repository name slash replacement

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 commented:
@nuclearsandwich, this is ready for re-review

nuclearsandwich commented on content/index.md:
```diff
@@ -1,18 +1,22 @@
 ---
-title: OSRA Infrastructure Project
+title: Open robotics infrastructure project
```
```suggestion
title: Open Robotics Infrastructure Project
```

Crola1702 replied:
Shouldn't be this "Open Robotics Infrastructure project"? Refering to the comment below

nuclearsandwich commented on Rules:
```diff
@@ -1,16 +1,18 @@
   @items.create('', { members: members }, '/team/index.html')
-  @items.create('', {projects: projects}, '/documentation/index.html')
+  @items.create('', { projects: projects }, '/documentation/index.html')
+  @items.create('', {}, '/repos/index.html')
```
Can we remove the preprocessor-created repos index by combining the `repos.erb` layout content into `content/repos/index.html`?

Crola1702 replied:
Ahh that's a good idea. Do you think it makes sense to do this exact same approach for projects and members?

Also, I'd like your opinion on naming it `repos/index.erb` instead. It's a personal preference, as for me it's ugly to find an html file with `<%= %>` in it :sweat_smile:

Crola1702 replied:
Solved in 5bc43c5

nuclearsandwich commented on content/index.md:
```diff
@@ -1,18 +1,22 @@
-title: OSRA Infrastructure Project
+title: Open robotics infrastructure project
 ---
+Welcome to the open robotics infrastructure project.
```
```suggestion
Welcome to the Open Robotics Infrastructure Project.
```

"Open Robotics" and "Infrastructure" are both proper nouns in this case.

nuclearsandwich reviewed (approved):
Some small nits but otherwise good to go!

Crola1702 committed:
Make index layouts actual html content. And remove preprocessor

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Apply suggestions from code review

Co-authored-by: Steven! Ragnarök <nuclearsandwich@users.noreply.github.com>
